vhost_user: Add support for SHMEM_MAP/UNMAP backend requests

mtjhrc opened the pull request:
### Summary of the PR

This PR adds support for backend requests `SHMEM_MAP` and `SHMEM_UNMAP`. 
This allows the device to memory map a file descriptor into a  `Shared Memory Region`.

The `SHMEM_MAP` and `SHMEM_UNMAP` are part of this QEMU RFC:
> [RFC PATCH v2 0/5] vhost-user: Add SHMEM_MAP/UNMAP requests
https://mail.gnu.org/archive/html/qemu-devel/2024-06/msg05736.html

I don't anticipate these specific requests to change much, the discussion for this RFC is mainly about a situation where one device is accessing shared memory mapped by another device.

These requests are required for implementing blob resource support in vhost-user-gpu. 
[The current gpu PR](
https://github.com/rust-vmm/vhost-device/pull/668
), doesn't implement blob resources and as such doesn't require this, but I am working on creating another PR with blob resource support.

This is based on a commit by Albert Esteve <aesteve@redhat.com> in his branch/fork.

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR have Signed-Off-By trailers (with
  `git commit -s`), and the commit message has max 60 characters for the
  summary and max 75 characters for each description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

stefano-garzarella commented on vhost/src/vhost_user/backend_req.rs:
```diff
@@ -183,6 +183,16 @@ impl VhostUserFrontendReqHandler for Backend {
         )
     }
+
+    /// Forward vhost-user memory map file requests to the backend.
```
```suggestion
    /// Forward vhost-user memory map file requests to the fronted.
```

stefano-garzarella commented on vhost/src/vhost_user/frontend_req_handler.rs:
```diff
@@ -52,6 +52,15 @@ pub trait VhostUserFrontendReqHandler {
+    /// Handle shared memory region unmapping requests.
+    fn shmem_unmap(&self, _req: &VhostUserMMap) -> HandlerResult<u64> {
+        Err(std::io::Error::from_raw_os_error(libc::ENOSYS))
+    }
```
I'd add a new line here

stefano-garzarella commented on vhost/src/vhost_user/message.rs:
```diff
@@ -194,6 +194,10 @@ enum_value! {
         SHARED_OBJECT_REMOVE = 7,
         /// Lookup for a virtio shared object.
         SHARED_OBJECT_LOOKUP = 8,
+        /// Map memory into guest address space ()
```
The last few characters ` ()`  may be typos.

stefano-garzarella commented on vhost/src/vhost_user/message.rs:
```diff
@@ -990,6 +994,48 @@ impl VhostUserMsgValidator for VhostUserTransferDeviceState {
+    /// Size of region to map.
+    pub len: u64,
+    /// Flags for the mmap operation
+    pub flags: VhostUserMMapFlags,
```
I'd follow what we did for other struct like `struct VhostUserMsgHeader` and `struct VhostUserHeaderFlag`. So here I'd use `u64`, also because in the `VhostUserMsgValidator` implementation you're already calling `.bits()`.

stefano-garzarella reviewed (commented):
Author and SoB are different, can you fix it, or add @aesteve-rh SoB if he is a co-author.

Please check also the coverage and add a line in the changelog.

mtjhrc commented:
I have addressed the feedback, you can take another look.

Also, should I add some comment that this feature is not really "stable"? Or will PR have to wait until it gets merged in QEMU?

mtjhrc commented:
The coverage was too low, so I wanted to add a test for the default impl of the frontend request handlers, but I also added coverage for all the methods, since that seemed quite trivial. This has increased the coverage too much, so I had to increase the coverage config value.

stefano-garzarella commented:
The code LGTM, but before merging this PR, I'd like to wait at least the spec merged in QEMU.
We recently removed some code that was never merged in the spec, so I'd like to avoid making the same mistake.
If the QEMU RFC still requires discussion about the code, maybe we can try to merge at least the changes to the specification so we can move forward here as well.

WDYT?

mtjhrc commented:
I understand the concern. I'm talking to @aesteve-rh to see if we can figure something out

alyssais commented on vhost/src/vhost_user/message.rs:
```diff
@@ -990,6 +994,48 @@ impl VhostUserMsgValidator for VhostUserTransferDeviceState {
+// Bit mask for flags in VhostUserBackendMapMsg backend message
+bitflags! {
+    #[derive(Copy, Clone, Debug, Eq, PartialEq, Default)]
+    /// Flags for virtio-fs backend messages.
```
virtio-fs?

mtjhrc replied:
Oops, fixed it. It was a leftover from the removed virtio-fs DAX messages.

dorindabassey commented on vhost/src/vhost_user/message.rs:
```diff
@@ -990,6 +994,48 @@ impl VhostUserMsgValidator for VhostUserTransferDeviceState {
+
+impl VhostUserMsgValidator for VhostUserMMap {
+    fn is_valid(&self) -> bool {
+        (self.flags & !VhostUserMMapFlags::all().bits()) == 0
+            && self.fd_offset.checked_add(self.len).is_some()
+            && self.shm_offset.checked_add(self.len).is_some()
+    }
```
do you want to add more checks relevant to the struct? e.g a check based on valid values for `shmid`?

mtjhrc replied:
From virtio spec:
> A device may have multiple shared memory regions associated with it. Each region has a shmid to identify it, the meaning of which is device-specific. 

According to the `virtio` (not vhost) spec the valid values of `shmid` are dependent on the device type (and the advertised device features), so I don't think you could check it here generally.

dorindabassey replied:
ok, fair enough for `shmid`, you can leave it as it is.

dorindabassey commented on vhost/src/vhost_user/frontend_req_handler.rs:
```diff
@@ -395,6 +437,16 @@ mod tests {
+
+        fn shmem_map(&mut self, req: &VhostUserMMap, _fd: &dyn AsRawFd) -> HandlerResult<u64> {
+            assert_eq!({ req.shmid }, 0);
+            Ok(!self.shmem_mappings.insert((req.shm_offset, req.len)) as u64)
```
i think it's better to remove the negation (!) and use a direct check to return meaningful values. e.g `if self.shmem_mappings.insert(req.shm_offset, req.len) { Ok(1) } else { Ok(0) }` and why the double braces? missing &?

mtjhrc replied:
By braces do you mean `( )`? The `(req.shm_offset, req.len)` is a tuple.

In regards to the negation with cast instead of explicit `if`, I considered to write it like you did, I mainly wrote it like this, because it is symmetric with the existing `shared_object_add`, `shared_object_remove` above:
```Rust
fn shared_object_add(&mut self, uuid: &VhostUserSharedMsg) -> HandlerResult<u64> {
    Ok(!self.shared_objects.insert(uuid.uuid) as u64)
}
fn shared_object_remove(&mut self, uuid: &VhostUserSharedMsg) -> HandlerResult<u64> {
    Ok(!self.shared_objects.remove(&uuid.uuid) as u64)
}
```

I am not sure if I should change it.

dorindabassey replied:
>  The (req.shm_offset, req.len) is a tuple.

ah right!

> because it is symmetric with the existing shared_object_add, shared_object_remove above:

doesn't necessarily mean it's the correct approach.

mtjhrc replied:
The `{ }` braces are required here to force the value to be copied. Without the braces this fails to compile because  `assert_eq!` macro tries to take a reference to `req.shmid` which is not possible. This is because the struct `VhostUserMMap` is `repr(packed)` so it is `unsafe` to take references to its fields (they may not be aligned).

dorindabassey commented on vhost/src/vhost_user/frontend_req_handler.rs:
```diff
@@ -395,6 +437,16 @@ mod tests {
+
+        fn shmem_unmap(&mut self, req: &VhostUserMMap) -> HandlerResult<u64> {
+            assert_eq!({ req.shmid }, 0);
+            Ok(!self.shmem_mappings.remove(&(req.shm_offset, req.len)) as u64)
```
same here

dorindabassey commented on vhost/src/vhost_user/backend_req.rs:
```diff
@@ -191,10 +191,16 @@ mod tests {
     use super::*;
 
+    fn frontend_backend_pair() -> (Endpoint<VhostUserMsgHeader<BackendReq>>, Backend) {
+        let (p1, p2) = UnixStream::pair().unwrap();
```
Instead of using .unwrap() do you want to handle the error here?

mtjhrc replied:
Well it's utility function only used inside tests, I don't think the error needs to be forwarded here. I think the only situation where this can fail is when maximum file descriptor limit is reached.

dorindabassey commented:
> The code LGTM, but before merging this PR, I'd like to wait at least the spec merged in QEMU. We recently removed some code that was never merged in the spec, so I'd like to avoid making the same mistake. If the QEMU RFC still requires discussion about the code, maybe we can try to merge at least the changes to the specification so we can move forward here as well.
> 
> WDYT?

are you talking about the vhost-user specification https://qemu-project.gitlab.io/qemu/interop/vhost-user.html#message-specification?  
If so, I think it make sense to atleast have just the spec merged, since the patches are still RFC it might take a while before it's merged.

stefano-garzarella commented:
> > The code LGTM, but before merging this PR, I'd like to wait at least the spec merged in QEMU. We recently removed some code that was never merged in the spec, so I'd like to avoid making the same mistake. If the QEMU RFC still requires discussion about the code, maybe we can try to merge at least the changes to the specification so we can move forward here as well.
> > WDYT?
> 
> are you talking about the vhost-user specification https://qemu-project.gitlab.io/qemu/interop/vhost-user.html#message-specification? 

@dorindabassey yep

> If so, I think it make sense to atleast have just the spec merged, since the patches are still RFC it might take a while before it's merged.

yeah, if it's possible, I'd go in that direction.
So, I'm marking this PR as draft to be sure we don't merge this till we have the spec changes merged in QEMU.

stefano-garzarella commented on vhost/src/vhost_user/message.rs:
```diff
@@ -987,6 +991,45 @@ impl VhostUserMsgValidator for VhostUserTransferDeviceState {
+    /// Flags specifying access permissions of memory mapping of a file
+    /// Use VhostUserMMapFlags::default() for a read-only mapping
+    pub struct VhostUserMMapFlags: u64 {
+        /// Read-write permission
```
nit: `/// Read-write permission (0: read-only, 1: read-write)`

stefano-garzarella replied:
ah, I thought was a bitfields.

Or maybe `/// Read-write permission (if not set, it's read-only)`, but I see we already have something in the struct doc, so really not a strong opinion on this, feel free to do nothing ;-)

stefano-garzarella commented on vhost/src/vhost_user/backend_req.rs:
```diff
@@ -174,6 +174,16 @@ impl VhostUserFrontendReqHandler for Backend {
+
+    /// Forward vhost-user memory map file request to the frontend.
+    fn shmem_map(&self, req: &VhostUserMMap, fd: &dyn AsRawFd) -> HandlerResult<u64> {
+        self.send_message(BackendReq::SHMEM_MAP, req, Some(&[fd.as_raw_fd()]))
```
In `shared_object_*()` functions above we check if the related feature is negotiated, should we do something similar also here?

mtjhrc replied:
Done

stefano-garzarella commented on vhost/src/vhost_user/message.rs:
```diff
@@ -987,6 +991,45 @@ impl VhostUserMsgValidator for VhostUserTransferDeviceState {
+// SAFETY: Safe because all fields of VhostUserMMap are POD.
+unsafe impl ByteValued for VhostUserMMap {}
+
+impl VhostUserMsgValidator for VhostUserMMap {
```
Should we add a test for this, as we do for other `VhostUserMsgValidator`  implementations (e.g. `check_user_memory()`, `check_user_memory_region()`, etc.)

stefano-garzarella commented on vhost/src/vhost_user/message.rs:
```diff
@@ -987,6 +991,45 @@ impl VhostUserMsgValidator for VhostUserTransferDeviceState {
+
+impl VhostUserMsgValidator for VhostUserMMap {
+    fn is_valid(&self) -> bool {
+        VhostUserMMapFlags::from_bits(self.flags).is_some()
```
Should we also check if `padding` is set to 0 like we do in `VhostUserMemory` ?
This is more for the future, where we can re-use the padding for something else. That said, in the spec I don't see any push on that :-(

epilys replied:
> Is self.len == 0 valid?

It's not mentioned in the spec text, but in the current patches in QEMU it's rejected:

```diff
+VirtioSharedMemoryMapping *virtio_shared_memory_mapping_new(uint8_t shmid,
+                                                            int fd,
+                                                            uint64_t fd_offset,
+                                                            uint64_t shm_offset,
+                                                            uint64_t len,
+                                                            bool allow_write)
+{
+    VirtioSharedMemoryMapping *mapping;
+    MemoryRegion *mr;
+    g_autoptr(GString) mr_name = g_string_new(NULL);
+    uint32_t ram_flags;
+    Error *local_err = NULL;
+
+    if (len == 0) {
+        error_report("Shared memory mapping size cannot be zero");
+        return NULL;
+    }
+
```

> Should we also check if padding is set to 0 like we do in VhostUserMemory ?

Sounds too strict IMHO... what if padding is filled with garbage bytes but the rest of the message is valid?

stefano-garzarella replied:
IMO in the spec we should say that padding must be 0, so in the future we can potentially use it for something else. At least, this is what we usually do in Linux uAPI.

aesteve-rh replied:
> IMO in the spec we should say that padding must be 0, so in the future we can potentially use it for something else. At least, this is what we usually do in Linux uAPI.

That's a good point. Honestly, I did not think about it. I will send a follow-up patch (I have a few minor changes piled already...) once it lands on QEMU to complete the spec with both points raised here.

Perhaps we can add the check for `len==0` here and wait for the padding check for when I can send the follow-up.

mtjhrc replied:
I added the check that verifies that `len` cannot be zero. I also added a comment that we don't validate the padding because it's currently not specified in the spec.
